File: DOACBorderApp/Sources/DOACBorderApp/FrameLayout.swift
import CoreGraphics

enum PageMode: Equatable {
    case free
    case a4
    case a5
    case custom
}

enum PageOrientation: Equatable {
    case auto      // match the source image's own aspect ratio
    case portrait
    case landscape
}

struct FrameLayout: Equatable {
    let scale: CGFloat
    let left: Int
    let top: Int
    let right: Int
    let bottom: Int
    let bottomRight: Int
    let canvasWidth: Int
    let canvasHeight: Int
    let holeWidth: Int
    let holeHeight: Int

    static func make(mode: PageMode, imageSize: CGSize, spec: TemplateSpec,
                      customSizeMM: (width: Double, height: Double) = (210, 297),
                      orientation: PageOrientation = .auto,
                      pct: CGFloat = 0.08, minPx: CGFloat = 60, dpi: CGFloat = 300) -> FrameLayout {
        let mm: (width: Double, height: Double)
        switch mode {
        case .free: return freeForm(imageSize: imageSize, spec: spec, pct: pct, minPx: minPx)
        case .a4: mm = (210, 297)
        case .a5: mm = (148, 210)
        case .custom: mm = customSizeMM
        }
        var mmW = mm.width, mmH = mm.height
        switch orientation {
        case .auto: if imageSize.width > imageSize.height { swap(&mmW, &mmH) }
        case .portrait: if mmW > mmH { swap(&mmW, &mmH) }
        case .landscape: if mmH > mmW { swap(&mmW, &mmH) }
        }
        let canvasWidth = Int((mmW / 25.4 * dpi).rounded())
        let canvasHeight = Int((mmH / 25.4 * dpi).rounded())

        let scale = max(pct * CGFloat(min(canvasWidth, canvasHeight)), minPx) / spec.left
        let left = Int((spec.left * scale).rounded())
        let top = Int((spec.top * scale).rounded())
        let right = Int((spec.right * scale).rounded())
        let bottom = Int((spec.bottom * scale).rounded())
        let bottomRight = Int((spec.bottomRight * scale).rounded())

        return FrameLayout(
            scale: scale, left: left, top: top, right: right, bottom: bottom, bottomRight: bottomRight,
            canvasWidth: canvasWidth, canvasHeight: canvasHeight,
            holeWidth: canvasWidth - left - right, holeHeight: canvasHeight - top - bottom
        )
    }

    static func freeForm(imageSize: CGSize, spec: TemplateSpec, pct: CGFloat = 0.08, minPx: CGFloat = 60) -> FrameLayout {
        let leftTarget = max(pct * min(imageSize.width, imageSize.height), minPx)
        let scale = leftTarget / spec.left
        let left = Int((spec.left * scale).rounded())
        let top = Int((spec.top * scale).rounded())
        let right = Int((spec.right * scale).rounded())
        let bottom = Int((spec.bottom * scale).rounded())
        let bottomRight = Int((spec.bottomRight * scale).rounded())
        let holeWidth = Int(imageSize.width.rounded())
        let holeHeight = Int(imageSize.height.rounded())
        return FrameLayout(
            scale: scale, left: left, top: top, right: right, bottom: bottom, bottomRight: bottomRight,
            canvasWidth: left + holeWidth + right, canvasHeight: top + holeHeight + bottom,
            holeWidth: holeWidth, holeHeight: holeHeight
        )
    }
}

File: web/js/frameLayout.js
const DEFAULT_CUSTOM_SIZE_MM = { width: 210, height: 297 };

export function makeLayout(mode, imageSize, spec, opts = {}) {
  const {
    customSizeMM = DEFAULT_CUSTOM_SIZE_MM,
    orientation = 'auto',
    pct = 0.08,
    minPx = 60,
    dpi = 300,
  } = opts;

  if (mode === 'free') {
    return freeForm(imageSize, spec, pct, minPx);
  }

  let mm;
  if (mode === 'a4') mm = { width: 210, height: 297 };
  else if (mode === 'a5') mm = { width: 148, height: 210 };
  else if (mode === 'custom') mm = customSizeMM;
  else throw new Error(`Unknown page mode: ${mode}`);

  let mmW = mm.width;
  let mmH = mm.height;
  if (orientation === 'auto') {
    if (imageSize.width > imageSize.height) [mmW, mmH] = [mmH, mmW];
  } else if (orientation === 'portrait') {
    if (mmW > mmH) [mmW, mmH] = [mmH, mmW];
  } else if (orientation === 'landscape') {
    if (mmH > mmW) [mmW, mmH] = [mmH, mmW];
  }

  const canvasWidth = Math.round((mmW / 25.4) * dpi);
  const canvasHeight = Math.round((mmH / 25.4) * dpi);

  const scale = Math.max(pct * Math.min(canvasWidth, canvasHeight), minPx) / spec.left;
  const left = Math.round(spec.left * scale);
  const top = Math.round(spec.top * scale);
  const right = Math.round(spec.right * scale);
  const bottom = Math.round(spec.bottom * scale);
  const bottomRight = Math.round(spec.bottomRight * scale);

  return {
    scale, left, top, right, bottom, bottomRight,
    canvasWidth, canvasHeight,
    holeWidth: canvasWidth - left - right,
    holeHeight: canvasHeight - top - bottom,
  };
}

function freeForm(imageSize, spec, pct, minPx) {
  const leftTarget = Math.max(pct * Math.min(imageSize.width, imageSize.height), minPx);
  const scale = leftTarget / spec.left;
  const left = Math.round(spec.left * scale);
  const top = Math.round(spec.top * scale);
  const right = Math.round(spec.right * scale);
  const bottom = Math.round(spec.bottom * scale);
  const bottomRight = Math.round(spec.bottomRight * scale);
  const holeWidth = Math.round(imageSize.width);
  const holeHeight = Math.round(imageSize.height);
  return {
    scale, left, top, right, bottom, bottomRight,
    canvasWidth: left + holeWidth + right,
    canvasHeight: top + holeHeight + bottom,
    holeWidth, holeHeight,
  };
}
